go/vt/mysqlctl/backupengine.go

```diff
@@ -50,3 +56,84 @@ func GetBackupEngine() (BackupEngine, error) {
 	}
 	return be, nil
 }
+
+func findBackupToRestore(ctx context.Context, cnf *Mycnf, mysqld MysqlDaemon, logger logutil.Logger, dir string, bhs []backupstorage.BackupHandle, bm interface{}) (backupstorage.BackupHandle, error) {
+	var bh backupstorage.BackupHandle
+	var index int
+
+	for index = len(bhs) - 1; index >= 0; index-- {
+		bh = bhs[index]
+		rc, err := bh.ReadFile(ctx, backupManifest)
+		if err != nil {
+			log.Warningf("Possibly incomplete backup %v in directory %v on BackupStorage: can't read MANIFEST: %v)", bh.Name(), dir, err)
+			continue
+		}
+
+		err = json.NewDecoder(rc).Decode(&bm)
+		rc.Close()
+		if err != nil {
+			log.Warningf("Possibly incomplete backup %v in directory %v on BackupStorage (cannot JSON decode MANIFEST: %v)", bh.Name(), dir, err)
+			continue
+		}
+
+		logger.Infof("Restore: found backup %v %v to restore", bh.Directory(), bh.Name())
+		break
+	}
+	if index < 0 {
+		// There is at least one attempted backup, but none could be read.
+		// This implies there is data we ought to have, so it's not safe to start
+		// up empty.
+		return nil, errors.New("backup(s) found but none could be read, unsafe to start up empty, restart to retry restore")
+	}
+
+	// shutdown mysqld if it is running
+	// if we are retrying a failed restore, it is possible that mysqld failed to start
+	// so check for that
+	waitCtx, cancel := context.WithTimeout(ctx, *topo.RemoteOperationTimeout)
+	defer cancel()
+	err := mysqld.Wait(waitCtx, cnf)
+	if err != nil || (waitCtx.Err() != nil && waitCtx.Err() == context.DeadlineExceeded) {
+		// do nothing
+	} else {
+		logger.Infof("Restore: shutdown mysqld")
+		// Starting from here we won't be able to recover if we get stopped by a cancelled
```

Comment:
yay!

Reply:
turns out we still need to use the background context. See #2310

Reply:
Isn't that issue just talking about the fact that the data dir will be incomplete, which is what this PR makes us now able to handle?

Reply:
It seems to go deeper than that, though I don't understand it completely. Empirically, the test starts failing if I change the context, even with the other changes I have made.

Reply:
I get it now. The test was testing to make sure that even if a grpc client dropped (hence canceling the context), the restore process should complete. This was necessary because `we won't be able to recover if we get stopped by a cancelled context`. Now that we are making it possible to retry restore even if it failed at some mid-point, the test can be changed to validate the behavior that scenario instead of expecting the restore to complete regardless.
@enisoc does that sound right?

Reply:
Yeah your diagnosis makes sense.

Reply:
@enisoc It just occurred to me to check all the places that were changed by #2310. It seems like other calls to mysqld.Start / Stop in this file can also be done with the regular context instead of `context.Background()`. What do you think?